### PR TITLE
[PNUDEV-43] Create feedback functionality

### LINKED ISSUE #43: Create feedback functionality
1. Add public page, where user will be able to send feedback and info how to contact him back (optionally, because we have to have anonymys feedback as well).
2. Add page to admin panel, where admin will be able to see feedbacks (sorted by date and with pagination). New feedbacks have to be hignlighted.
3. Add menu item with the link to the feedback page to the header of admin panel. This menu item should show nunber of new (not viewed) feedbacks e.g. (+1), if there was new feedback.
4. Keep XSS in mind as long as we request input from user, that will be stored to db and shown on the admin panel.

### PR DESCRIPTION
**Pull request by volodymyrvynnyk**
image
image
image
image
image

### CODE REVIEW

File: src/main/java/com/pnu/dev/shpaltaif/controller/FeedbackController.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    @GetMapping
+    public String createForm() {
+        return "feedback/form";
+    }
+
+    @PostMapping("/create")
+    public String create(Feedback feedback, Model model) {
+        feedbackService.create(feedback);
+        model.addAttribute("feedback", feedback);
+        return "feedback/success";
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/controller/FeedbackController.java]: We shouldn't render the page after POST.

Check how it's handled here:
https://github.com/PNUdev/pnu-feedback/blob/master/src/main/java/com/pnu/dev/pnufeedback/controller/FeedbackSubmissionController.java#L124-L125

File: src/main/resources/application.yaml [hunk 2]
```diff
@@ -20,6 +20,7 @@ spring:
           cache:
             missing_cache_strategy: create
     hibernate.ddl-auto: validate
+    open-in-view: false
```
Comment (Igor-Polatajko) [inline src/main/resources/application.yaml:23]: It's actually good, to set it to false...
But now you have to perform manual regression testing of the application to make sure that we don't rely on this feature anywhere.

File: src/main/resources/templates/admin/feedback/index.ftl [hunk 3]
```diff
@@ -0,0 +1,57 @@
+<#include "../include/header.ftl">
+<h2 class="text-center my-3">Відгуки</h2>
+
+<#if !feedbacks?has_content >
+    <h3 class="text-center">Список відкритих відповідей пустий</h3>
+<#else>
+    <table class="table table-striped mx-3">
+        <tbody>
+        <#list feedbacks.getContent() as feedback >
```
Comment (Igor-Polatajko) [inline src/main/resources/templates/admin/feedback/index.ftl]: Why not `feedbacks.content`?

File: src/main/resources/templates/include/footer.ftl [hunk 4]
```diff
@@ -1,3 +1,4 @@
 <h1>Footer</h1>
+<a href="/feedbacks">Залишити відгук</a>
```
Comment (Igor-Polatajko) [inline src/main/resources/templates/include/footer.ftl]: Зворотній зв'язок

File: src/main/resources/templates/telegram/start.ftl [hunk 5]
```diff
@@ -1,3 +1,4 @@
 <b>Вітаємо, Shpalta IF бот повідомлятиме Вас про нові пости на ${appBasePath}</b>
 
-Категорії, сповіщення з яких Ви отримуватимете, можна налаштувати, натиснувши на кнопку <b>"Налаштувати сповіщення"</b> або відправивши /settings
\ No newline at end of file
+Категорії, сповіщення з яких Ви отримуватимете, можна налаштувати, натиснувши на кнопку <b>"Налаштувати
```
Comment (Igor-Polatajko) [inline src/main/resources/templates/telegram/start.ftl]: Why changed formatting?

File: src/test/java/com/pnu/dev/shpaltaif/integration/service/FeedbackServiceIntegrationTest.java [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    private FeedbackRepository feedbackRepository;
+
+    @Test
+    public void createThanReviewThanDeactivateTest() {
+
+        assertEquals(0, feedbackService.countUnreviewed());
+        assertEquals(0, feedbackService.findAll(PAGEABLE).getTotalElements());
+
+        //Create feedback
+        Feedback feedbackToCreate = Feedback.builder()
```
Comment (Igor-Polatajko) [inline src/test/java/com/pnu/dev/shpaltaif/integration/service/FeedbackServiceIntegrationTest.java:43]: We agreed to perform each integration test in transaction, but set up data in the first one, that wouldn't rollback immediately.

Just to follow A-A-A pattern.

But let it be like this for now. Maybe, one day, this test will be rewritten.

File: src/main/resources/templates/feedback/form.ftl [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<#include "../include/header.ftl" >
+<form method="POST" action="/feedbacks/create">
+    <input type="hidden" name="${_csrf.parameterName}" value="${_csrf.token}"/>
+    Що ви хотіли б запропонувати покращити?
```
Comment (Igor-Polatajko) [inline src/main/resources/templates/feedback/form.ftl]: Що ви хотіли б запропонувати чи покращити?

File: src/main/java/com/pnu/dev/shpaltaif/service/FeedbackServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,90 @@
...
+                .reviewed(true)
+                .build()
+        ).collect(Collectors.toList());
+
+        feedbackRepository.saveAll(reviewedFeedbacks);
+    }
+
+
+    private Feedback findById(Long id) {
+        return feedbackRepository.findById(id)
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/service/FeedbackServiceImpl.java]: It's used only in one place, so no need it private method

File: src/main/java/com/pnu/dev/shpaltaif/service/FeedbackServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+                .createdAt(feedbackFromDb.getCreatedAt())
+                .reviewed(feedbackFromDb.isReviewed())
+                .active(false)
+                .build();
+
+        feedbackRepository.save(reviewedFeedback);
+    }
+
+    private void markAllAsReviewed(List<Feedback> feedbacks) {
+        List<Feedback> reviewedFeedbacks = feedbacks.stream().map(feedback -> Feedback.builder()
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/service/FeedbackServiceImpl.java]: No need to rebuild object. Just use `toBuilder()`

File: src/main/java/com/pnu/dev/shpaltaif/service/FeedbackServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+                .build();
+
+        feedbackRepository.save(feedbackToCreate);
+    }
+
+    @Override
+    public void deactivate(Long id) {
+
+        Feedback feedbackFromDb = findById(id);
+        Feedback reviewedFeedback = Feedback.builder()
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/service/FeedbackServiceImpl.java]: same here

File: src/main/resources/templates/telegram/start.ftl (hunk 5 above)
Comment (volodymyrvynnyk) [reply src/main/resources/templates/telegram/start.ftl]: Sorry, it has been changed by autoformat

File: src/main/resources/application.yaml (hunk 2 above)
Comment (volodymyrvynnyk) [reply src/main/resources/application.yaml:23]: Yes, I tested the app manually, I didn't find bugs so far

File: src/main/java/com/pnu/dev/shpaltaif/controller/FeedbackController.java [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+
+    @GetMapping("/after-submit")
+    public String showAfterSubmitPage() {
+
+        return "feedback/success";
+    }
+
+    @PostMapping("/create")
+    public String create(Feedback feedback) {
+        feedbackService.create(feedback);
```
Comment (Igor-Polatajko) [inline src/main/java/com/pnu/dev/shpaltaif/controller/FeedbackController.java:41]: use flash attribute to show next page only once

**Review by Igor-Polatajko — APPROVED** [review]
Looks good, one comment

**Pull request merged**